calling `system.card[T](x: set[T])` with `T` of `int8` or `uint8` uses mismatched C array sizes

### Description

When calling `system.card[T](x: set[T])` on either `set[int8]` or `set[uint8]`, Nim generates a function
```c
static N_INLINE(NI, cardSet)(tyArray__S8ib5N31oJbAkqOZTlNg5Q s, NI len);
...
static N_INLINE(NI, cardSet)(tyArray__S8ib5N31oJbAkqOZTlNg5Q s, NI len) {
    ...
}
```

which is called from
```c
N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
{
	tySet_tyInt8__chXunH2dwinSkhpA6JnsXw T1_;
	nimfr_("cmdfile", "cmdfile.nim");
	nimln_(1, "cmdfile.nim");
	nimZeroMem((void*)T1_, sizeof(tySet_tyInt8__chXunH2dwinSkhpA6JnsXw));
	(void)(cardSet(T1_, 32));
	nimTestErrorFlag();
	popFrame();
}
}
```

where `tyArray__S8ib5N31oJbAkqOZTlNg5Q` and `tySet_tyInt8__chXunH2dwinSkhpA6JnsXw` are defined as
```c
typedef NU8 tySet_tyInt8__chXunH2dwinSkhpA6JnsXw[32];
typedef NU8 tyArray__S8ib5N31oJbAkqOZTlNg5Q[8192];
```

It's calling (through/via a `void *`, which in itself is fine) a function expecting an `NU8[8192]` with an `NU8[32]`. It doesn't ever access these extra (8192-32) elements of the array, because it sets `cardSet`'s `len` parameter to 32, but given that the sizes are both known ahead of time, it's feasible to match them.

### Nim Version

```
Nim Compiler Version 1.6.10 [Linux: amd64]
Compiled at 2022-11-23
Copyright (c) 2006-2021 by Andreas Rumpf
```

```
Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-12-02
Copyright (c) 2006-2022 by Andreas Rumpf
```

```
$ gcc --version
gcc (Debian 12.2.0-9) 12.2.0
```

### Current Output

```text
$ nim c --passC=-flto --passL=-flto --eval:"discard card(default(set[uint8]))"
Hint: used config file 'Nim/config/nim.cfg' [Conf]
Hint: used config file 'Nim/config/config.nims' [Conf]
...................................................................
CC: cmdfile.nim
Hint:  [Link]
.cache/nim/cmdfile_d/@mcmdfile.nim.c: In function ‘NimMainModule’:
.cache/nim/cmdfile_d/@mcmdfile.nim.c:240:16: warning: ‘cardSet’ accessing 8192 bytes in a region of size 32 [-Wstringop-overflow=]
  240 |         (void)(cardSet(T1_, 32));
      |                ^
.cache/nim/cmdfile_d/@mcmdfile.nim.c:240:16: note: referencing argument 1 of type ‘NU8[8192]’
.cache/nim/cmdfile_d/@mcmdfile.nim.c:153:8: note: in a call to function ‘cardSet’
  153 | static N_INLINE(NI, cardSet)(tyArray__S8ib5N31oJbAkqOZTlNg5Q s, NI len) {
      |        ^
Hint: mm: orc; threads: on; opt: none (DEBUG BUILD, `-d:release` generates faster code)
27506 lines; 0.577s; 30.328MiB peakmem; proj: cmdfile; out: .cache/nim/cmdfile_d/cmdfile [SuccessX]
```


### Expected Output

```text
No usage of mismatched array sizes on caller/callee side.
```


### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -2122,7 +2122,7 @@ proc genSetOp(p: BProc, e: PNode, d: var TLoc, op: TMagic) =
     of mCard:
       var a: TLoc
       initLocExpr(p, e[1], a)
-      putIntoDest(p, d, e, ropecg(p.module, "#cardSet($1, $2)", [rdCharLoc(a), size]))
+      putIntoDest(p, d, e, ropecg(p.module, "#cardSet($1, $2)", [addrLoc(p.config, a), size]))
     of mLtSet, mLeSet:
       getTemp(p, getSysType(p.module.g.graph, unknownLineInfo, tyInt), i) # our counter
       initLocExpr(p, e[1], a)

--- a/lib/system/sets.nim
+++ b/lib/system/sets.nim
@@ -10,11 +10,20 @@
 # set handling
 
 type
-  NimSet8 = array[0..32-1, uint8]
-  NimSet16 = array[0..8192-1, uint8]
+  NimSet16 = array[0..16-1, uint8]
+  NimSet32 = array[0..32-1, uint8]
+  NimSet64 = array[0..64-1, uint8]
+  NimSet128 = array[0..128-1, uint8]
+  NimSet256 = array[0..256-1, uint8]
+  NimSet512 = array[0..512-1, uint8]
+  NimSet1024 = array[0..1024-1, uint8]
+  NimSet2048 = array[0..2048-1, uint8]
+  NimSet4096 = array[0..4096-1, uint8]
+  NimSet8192 = array[0..8192-1, uint8]
+  NimSet = NimSet16 | NimSet32 | NimSet64 | NimSet128 | NimSet256 |
+    NimSet512 | NimSet1024 | NimSet2048 | NimSet4096 | NimSet8192
 
-
-proc cardSetImpl(s: NimSet8 | NimSet16, len: int): int {.inline.} =
+proc cardSetImpl(s: NimSet, len: int): int {.inline.} =
   var i = 0
   result = 0
   when defined(x86) or defined(amd64):
@@ -26,8 +35,17 @@ proc cardSetImpl(s: NimSet8 | NimSet16, len: int): int {.inline.} =
     inc(result, countBits32(uint32(s[i])))
     inc(i, 1)
 
-proc cardSet(s: NimSet8, len: int): int {.compilerproc, inline.} =
-  result = cardSetImpl(s, len)
+template cardSetDef(t) = 
+  proc cardSet(s: t, len: int): int {.compilerproc, inline.} =
+    result = cardSetImpl(s, len)
 
-proc cardSet(s: NimSet16, len: int): int {.compilerproc, inline.} =
-  result = cardSetImpl(s, len)
+cardSetDef(NimSet16)
+cardSetDef(NimSet32)
+cardSetDef(NimSet64)
+cardSetDef(NimSet128)
+cardSetDef(NimSet256)
+cardSetDef(NimSet512)
+cardSetDef(NimSet1024)
+cardSetDef(NimSet2048)
+cardSetDef(NimSet4096)
+cardSetDef(NimSet8192)

--- a/lib/system/sets.nim
+++ b/lib/system/sets.nim
@@ -36,7 +36,7 @@ proc cardSetImpl(s: NimSet, len: int): int {.inline.} =
     inc(i, 1)
 
 template cardSetDef(t) = 
-  proc cardSet(s: t, len: int): int {.compilerproc, inline.} =
+  proc `cardSet t`(s: t, len: int): int {.inject, inline.} =
     result = cardSetImpl(s, len)
 
 cardSetDef(NimSet16)
@@ -49,3 +49,28 @@ cardSetDef(NimSet1024)
 cardSetDef(NimSet2048)
 cardSetDef(NimSet4096)
 cardSetDef(NimSet8192)
+
+proc cardSet(s: NimSet8192, len: int): int {.compilerproc, inline.} =
+  case len
+  of 16:
+    result = cardSetNimSet16(cast[ptr NimSet16](s.addr)[], len)
+  of 32:
+    result = cardSetNimSet32(cast[ptr NimSet32](s.addr)[], len)
+  of 64:
+    result = cardSetNimSet64(cast[ptr NimSet64](s.addr)[], len)
+  of 128:
+    result = cardSetNimSet128(cast[ptr NimSet128](s.addr)[], len)
+  of 256:
+    result = cardSetNimSet256(cast[ptr NimSet256](s.addr)[], len)
+  of 512:
+    result = cardSetNimSet512(cast[ptr NimSet512](s.addr)[], len)
+  of 1024:
+    result = cardSetNimSet1024(cast[ptr NimSet1024](s.addr)[], len)
+  of 2048:
+    result = cardSetNimSet2048(cast[ptr NimSet2048](s.addr)[], len)
+  of 4096:
+    result = cardSetNimSet4096(cast[ptr NimSet4096](s.addr)[], len)
+  of 8192:
+    result = cardSetNimSet8192(cast[ptr NimSet8192](s.addr)[], len)
+  else:
+    discard

--- a/lib/system/sets.nim
+++ b/lib/system/sets.nim
@@ -10,10 +10,11 @@
 # set handling
 
 type
-  NimSet = array[0..4*2048-1, uint8]
+  NimSet8 = array[0..32-1, uint8]
+  NimSet16 = array[0..8192-1, uint8]
 
 
-proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
+proc cardSetImpl(s: NimSet8 | NimSet16, len: int): int {.inline.} =
   var i = 0
   result = 0
   when defined(x86) or defined(amd64):
@@ -24,3 +25,9 @@ proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
   while i < len:
     inc(result, countBits32(uint32(s[i])))
     inc(i, 1)
+
+proc cardSet(s: NimSet8, len: int): int {.compilerproc, inline.} =
+  result = cardSetImpl(s, len)
+
+proc cardSet(s: NimSet16, len: int): int {.compilerproc, inline.} =
+  result = cardSetImpl(s, len)

--- a/lib/system/sets.nim
+++ b/lib/system/sets.nim
@@ -10,20 +10,9 @@
 # set handling
 
 type
-  NimSet16 = array[0..16-1, uint8]
-  NimSet32 = array[0..32-1, uint8]
-  NimSet64 = array[0..64-1, uint8]
-  NimSet128 = array[0..128-1, uint8]
-  NimSet256 = array[0..256-1, uint8]
-  NimSet512 = array[0..512-1, uint8]
-  NimSet1024 = array[0..1024-1, uint8]
-  NimSet2048 = array[0..2048-1, uint8]
-  NimSet4096 = array[0..4096-1, uint8]
-  NimSet8192 = array[0..8192-1, uint8]
-  NimSet = NimSet16 | NimSet32 | NimSet64 | NimSet128 | NimSet256 |
-    NimSet512 | NimSet1024 | NimSet2048 | NimSet4096 | NimSet8192
+  NimSet = array[0..8192-1, uint8]
 
-proc cardSetImpl(s: NimSet, len: int): int {.inline.} =
+proc cardSetImpl(s: openArray[uint8], len: int): int {.inline.} =
   var i = 0
   result = 0
   when defined(x86) or defined(amd64):
@@ -35,42 +24,5 @@ proc cardSetImpl(s: NimSet, len: int): int {.inline.} =
     inc(result, countBits32(uint32(s[i])))
     inc(i, 1)
 
-template cardSetDef(t) = 
-  proc `cardSet t`(s: t, len: int): int {.inject, inline.} =
-    result = cardSetImpl(s, len)
-
-cardSetDef(NimSet16)
-cardSetDef(NimSet32)
-cardSetDef(NimSet64)
-cardSetDef(NimSet128)
-cardSetDef(NimSet256)
-cardSetDef(NimSet512)
-cardSetDef(NimSet1024)
-cardSetDef(NimSet2048)
-cardSetDef(NimSet4096)
-cardSetDef(NimSet8192)
-
-proc cardSet(s: NimSet8192, len: int): int {.compilerproc, inline.} =
-  case len
-  of 16:
-    result = cardSetNimSet16(cast[ptr NimSet16](s.addr)[], len)
-  of 32:
-    result = cardSetNimSet32(cast[ptr NimSet32](s.addr)[], len)
-  of 64:
-    result = cardSetNimSet64(cast[ptr NimSet64](s.addr)[], len)
-  of 128:
-    result = cardSetNimSet128(cast[ptr NimSet128](s.addr)[], len)
-  of 256:
-    result = cardSetNimSet256(cast[ptr NimSet256](s.addr)[], len)
-  of 512:
-    result = cardSetNimSet512(cast[ptr NimSet512](s.addr)[], len)
-  of 1024:
-    result = cardSetNimSet1024(cast[ptr NimSet1024](s.addr)[], len)
-  of 2048:
-    result = cardSetNimSet2048(cast[ptr NimSet2048](s.addr)[], len)
-  of 4096:
-    result = cardSetNimSet4096(cast[ptr NimSet4096](s.addr)[], len)
-  of 8192:
-    result = cardSetNimSet8192(cast[ptr NimSet8192](s.addr)[], len)
-  else:
-    discard
+proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
+  result = cardSetImpl(s, len)

--- a/tests/sets/t20997.nim
+++ b/tests/sets/t20997.nim
@@ -1,6 +1,14 @@
 {.passC: "-flto".}
 {.passL: "-flto".}
 
-discard card(default(set[int8]))
-
-discard card(default(set[int16]))
+template f(n: int) = discard card(default(set[range[0 .. (1 shl n) - 1]]))
+f( 7)
+f( 8)
+f( 9)
+f(10)
+f(11)
+f(12)
+f(13)
+f(14)
+f(15)
+f(16)

--- a/tests/sets/t20997.nim
+++ b/tests/sets/t20997.nim
@@ -1,3 +1,7 @@
+discard """
+  joinable: false
+"""
+
 {.passC: "-flto".}
 {.passL: "-flto".}
 

--- a/tests/sets/t20997.nim
+++ b/tests/sets/t20997.nim
@@ -1,0 +1,6 @@
+{.passC: "-flto".}
+{.passL: "-flto".}
+
+discard card(default(set[int8]))
+
+discard card(default(set[int16]))

